### ISSUE TITLE
Return a promise from setState

### ISSUE THREAD

**Issue by anthonyshort** [segmentio/deku#41]
This shouldn't be too hard. Just return a promise and resolve it on an event.

**Comment by anthonyshort** [comment]
I think we might just remove the callback from `setState`. Almost every use-case can be covered in `afterUpdate`. I can't think of anything that would benefit from it.

**Comment by yoshuawuyts** [comment]
agreed. if `afterUpdate` is the only thing that's triggered, well, after an update it makes for more transparent code. That also means there's no need to return a `Promise`.

**Comment by ivan-kleshnin** [comment]
@anthonyshort did you consider to make `setState` actually synchronous?
I mean to make it represent *pending* state and not *rendered* state.
It was discussed in React community many times https://groups.google.com/forum/#!msg/reactjs/R61kPjs-yXE/bk5AGv78RYAJ
but got stucked because of backward compatibility issues.

**Comment by anthonyshort** [comment]
@ivan-kleshnin We're definitely more free to make changes like this. In general I don't like the hooks, and I've found we rarely use them anyway. The only one I can see being useful would be something like `afterRender` so you could update some canvas element using the latest state. 

I'm not sure I'm following the exact reasoning behind it yet though. What are the benefits to it? We don't access state on `this` so it probably wouldn't update the `state` object that gets passed in anyway.

**Comment by ivan-kleshnin** [comment]
> I'm not sure I'm following the exact reasoning behind it yet though. What are the benefits to it? We don't access state on this so it probably wouldn't update the state object that gets passed in anyway.

The main benefit is that code is synchronous. `this.state` is always the "latest" state. It's obviously easier to work with sync code, than with async. No callbacks. No promises. Easier to reason about it. You can `setState()` multiple time which is often required when we refactor business logic into functions. No rigid hooks.

Now you can fire `setState` once, then you can analize it in `afterRender` once. 
What if we want to change something after analysis? No way, right?!

Will you excuse me, if I just give you a link to a thread where we had a lot of dissussion on this subject
https://github.com/Yomguithereal/baobab/issues/190
Don't want to copy-past all that here. There are more links to more discussions on this subject from different people.

Basically, this very change was discussed and successfully implemented in the Baobab library and everyone seems to be happy about it. The context was slightly different, but I'm pretty sure this very argumentation applies to Deku as well.
